wrangler-transform/src/main/java/io/cdap/wrangler/WranglerErrorUtil.java

```diff
@@ -0,0 +1,125 @@
+import io.cdap.wrangler.api.DirectiveParseException;
+import io.cdap.wrangler.api.RecipeException;
+import io.cdap.wrangler.expression.ELException;
+import io.cdap.wrangler.utils.RecordConvertorException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Error util file to handle exceptions caught in Wrangler plugin
+ */
+public final class WranglerErrorUtil {
+
+  private static final Map<Class<? extends Throwable>, String> TERMINAL_EXCEPTIONS;
+
+  static {
+    TERMINAL_EXCEPTIONS = new HashMap<>();
+    TERMINAL_EXCEPTIONS.put(DirectiveParseException.class, "Parsing-Directive");
+    TERMINAL_EXCEPTIONS.put(DirectiveExecutionException.class, "Executing-Directive");
+    TERMINAL_EXCEPTIONS.put(DirectiveLoadException.class, "Loading-Directive");
+    TERMINAL_EXCEPTIONS.put(DirectiveNotFoundException.class, "NotFound-Directive");
+    TERMINAL_EXCEPTIONS.put(RecordConvertorException.class, "RecordConvertor");
+    TERMINAL_EXCEPTIONS.put(ELException.class, "ExpressionLanguage-Parsing");
+  }
+
+  /**
+   * Private constructor to prevent instantiation of this utility class.
+   * <p>
+   * This class is designed to contain only static utility methods for handling exceptions and
+   * should not be instantiated. Any attempt to create an instance of this class will result in an
+   * {@link IllegalStateException}.
+   */
+  private WranglerErrorUtil() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  /**
+   * Traverses the causal chain of the given Throwable to find specific exceptions. If a terminal
+   * exception is found, it returns a corresponding ProgramFailureException. Otherwise, a generic
+   * ProgramFailureException is returned.
+   *
+   * @param e            the Throwable to analyze
+   * @param errorMessage default error message if no terminal exception is found
+   * @param errorType    the error type to categorize the failure
+   * @return a ProgramFailureException with specific or generic error details
+   */
+  public static ProgramFailureException getProgramFailureExceptionDetailsFromChain(Throwable e,
+      String errorMessage, ErrorType errorType) {
+    if (e == null) {
+      return ErrorUtils.getProgramFailureException(
+          new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN), errorMessage, errorMessage,
+          errorType, false, null);
+    }
+
+    List<Throwable> causalChain = Throwables.getCausalChain(e);
+    RecipeException recipeException = null;
+
+    for (Throwable t : causalChain) {
+      if (t instanceof ProgramFailureException) {
+        return null; // Avoid multiple wrap
+      }
+      if (t instanceof PreconditionException) {
+        return getProgramFailureException((PreconditionException) t, errorType, "Precondition");
+      }
+      if (t instanceof RecipeException) {
+        // Store RecipeException as fallback, if there is only Recipe exception in casual chain
+        recipeException = (RecipeException) t;
```

Comment:
can we also create a list of `non-terminal_exceptions` with category to make it extensible?